server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentTracker.java

```diff
@@ -0,0 +1,425 @@
+     */
+    private volatile long seqNoLag;
+
+    /**
+     * Keeps the time (ms) lag computed so that we do not compute it for every request.
+     */
+    private volatile long timeMsLag;
+
+    /**
+     * Cumulative sum of size in bytes of segment files for which upload has started during remote refresh.
+     */
+    private volatile long uploadBytesStarted;
+
+    /**
+     * Cumulative sum of size in bytes of segment files for which upload has failed during remote refresh.
+     */
+    private volatile long uploadBytesFailed;
+
+    /**
+     * Cumulative sum of size in bytes of segment files for which upload has succeeded during remote refresh.
+     */
+    private volatile long uploadBytesSucceeded;
+
+    /**
+     * Cumulative sum of count of remote refreshes that have started.
+     */
+    private volatile long totalUploadsStarted;
+
+    /**
+     * Cumulative sum of count of remote refreshes that have failed.
+     */
+    private volatile long totalUploadsFailed;
+
+    /**
+     * Cumulative sum of count of remote refreshes that have succeeded.
+     */
+    private volatile long totalUploadsSucceeded;
+
+    /**
+     * Cumulative sum of rejection counts for this shard.
+     */
+    private final AtomicLong rejectionCount = new AtomicLong();
+
+    /**
+     * Keeps track of rejection count with each rejection reason.
+     */
+    private final Map<String, AtomicLong> rejectionCountMap = ConcurrentCollections.newConcurrentMap();
+
+    /**
+     * Map of name to size of the segment files created as part of the most recent refresh.
+     */
+    private volatile Map<String, Long> latestLocalFileNameLengthMap;
+
+    /**
+     * Set of names of segment files that were uploaded as part of the most recent remote refresh.
+     */
+    private final Set<String> latestUploadFiles = new HashSet<>();
+
+    /**
+     * Keeps the bytes lag computed so that we do not compute it for every request.
+     */
+    private volatile long bytesLag;
+
+    /**
+     * Holds count of consecutive failures until last success. Gets reset to zero if there is a success.
+     */
+    private final Streak failures = new Streak();
+
+    /**
+     * Provides moving average over the last N total size in bytes of segment files uploaded as part of remote refresh.
+     * N is window size. Wrapped with {@code AtomicReference} for dynamic changes in window size.
+     */
+    private final AtomicReference<MovingAverage> uploadBytesMovingAverageReference;
+
+    /**
+     * This lock object is used for making sure we do not miss any data
+     */
+    private final Object uploadBytesMutex = new Object();
+
+    /**
+     * Provides moving average over the last N upload speed (in bytes/s) of segment files uploaded as part of remote refresh.
+     * N is window size. Wrapped with {@code AtomicReference} for dynamic changes in window size.
+     */
+    private final AtomicReference<MovingAverage> uploadBytesPerSecMovingAverageReference;
```

Comment:
I think this can just be a volatile reference since you're only doing plain gets and sets.

Reply:
There are cases when the window size needs to be updated dynamically in which the reference is updated by having external synchronisation. For that reason, I have used AtomicReference so that when we want to get the underlying reference, it is always correct.